File: src/main/java/no/nav/syfo/kafka/mapper/SykepengesoknadDtoToSykepengesoknadMapper.java
package no.nav.syfo.kafka.mapper;

import no.nav.syfo.domain.dto.*;
import no.nav.syfo.kafka.sykepengesoknad.dto.SoknadsperiodeDTO;
import no.nav.syfo.kafka.sykepengesoknad.dto.SporsmalDTO;
import no.nav.syfo.kafka.sykepengesoknad.dto.SvarDTO;
import no.nav.syfo.kafka.sykepengesoknad.dto.SykepengesoknadDTO;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.Optional.ofNullable;

public final class SykepengesoknadDtoToSykepengesoknadMapper {

    private static <T extends Enum<T>, U extends Enum<U>> T konverter(Class<T> tClass, U from) {
        return ofNullable(from).map(n -> Enum.valueOf(tClass, n.name())).orElse(null);
    }

    private static Svar konverter(SvarDTO svar) {
        return Svar.builder()
                .verdi(svar.getVerdi())
                .build();
    }

    private static Sporsmal konverter(SporsmalDTO sporsmal) {
        return Sporsmal.builder()
                .id(sporsmal.getId())
                .tag(sporsmal.getTag())
                .sporsmalstekst(sporsmal.getSporsmalstekst())
                .undertekst(sporsmal.getUndertekst())
                .svartype(konverter(Svartype.class, sporsmal.getSvartype()))
                .min(sporsmal.getMin())
                .max(sporsmal.getMax())
                .kriterieForVisningAvUndersporsmal(konverter(Visningskriterie.class, sporsmal.getKriteriumForVisningAvUndersporsmal()))
                .svar(stream(sporsmal.getSvar())
                        .map(SykepengesoknadDtoToSykepengesoknadMapper::konverter)
                        .collect(Collectors.toList()))
                .undersporsmal(stream(sporsmal.getUndersporsmal())
                        .map(SykepengesoknadDtoToSykepengesoknadMapper::konverter)
                        .collect(Collectors.toList()))
                .build();
    }

    private static SoknadPeriode konverter(SoknadsperiodeDTO soknadPeriode) {
        return SoknadPeriode.builder()
                .fom(soknadPeriode.getFom())
                .tom(soknadPeriode.getTom())
                .grad(soknadPeriode.getSykmeldingsgrad())
                .build();
    }

    public static Sykepengesoknad konverter(SykepengesoknadDTO sykepengesoknad) {
        return Sykepengesoknad.builder()
                .id(sykepengesoknad.getId())
                .sykmeldingId(sykepengesoknad.getSykmeldingId())
                .aktorId(sykepengesoknad.getAktorId())
                .soknadstype(konverter(Soknadstype.class, sykepengesoknad.getType()))
                .status(sykepengesoknad.getStatus().name())
                .fom(sykepengesoknad.getFom())
                .tom(sykepengesoknad.getTom())
                .opprettet(sykepengesoknad.getOpprettet())
                .sendtNav(sykepengesoknad.getSendtNav())
                .sendtArbeidsgiver(sykepengesoknad.getSendtArbeidsgiver())
                .arbeidsgiver(sykepengesoknad.getArbeidsgiver().getNavn())
                .arbeidssituasjon(konverter(Arbeidssituasjon.class, sykepengesoknad.getArbeidssituasjon()))
                .startSykeforlop(sykepengesoknad.getStartSyketilfelle())
                .sykmeldingSkrevet(sykepengesoknad.getSykmeldingSkrevet())
                .korrigertAv(sykepengesoknad.getKorrigertAv())
                .korrigerer(sykepengesoknad.getKorrigerer())
                .soknadPerioder(stream(sykepengesoknad.getSoknadsperioder())
                        .map(SykepengesoknadDtoToSykepengesoknadMapper::konverter)
                        .collect(Collectors.toList()))
                .sporsmal(stream(sykepengesoknad.getSporsmal())
                        .map(SykepengesoknadDtoToSykepengesoknadMapper::konverter)
                        .collect(Collectors.toList()))
                .build();
    }

    private static <T> Stream<T> stream(List<T> list) {
        return list == null ? Stream.empty() : list.stream();
    }
}

File: src/main/kotlin/no/nav/syfo/kafka/mapper/SykepengesoknadDtoToSykepengesoknadMapper.kt
package no.nav.syfo.kafka.mapper

import no.nav.syfo.domain.dto.*
import no.nav.syfo.kafka.felles.DeprecatedSykepengesoknadDTO
import no.nav.syfo.kafka.felles.SoknadsperiodeDTO
import no.nav.syfo.kafka.felles.SporsmalDTO
import no.nav.syfo.kafka.felles.SvarDTO
import no.nav.syfo.kafka.felles.SykepengesoknadDTO

private inline fun <T : Enum<*>, reified U : Enum<*>> T?.enumValueOrNull(): U? =
    U::class.java.enumConstants.firstOrNull { it.name == this?.name }

private fun SvarDTO.toSvar(): Svar =
    Svar(verdi)

private fun SporsmalDTO.toSporsmal(): Sporsmal =
    Sporsmal(
        id = id!!,
        tag = tag!!,
        sporsmalstekst = sporsmalstekst,
        undertekst = undertekst,
        svartype = svartype.enumValueOrNull(),
        min = min,
        max = max,
        kriterieForVisningAvUndersporsmal = kriterieForVisningAvUndersporsmal.enumValueOrNull(),
        svar = svar!!.map { it.toSvar() },
        undersporsmal = undersporsmal?.map { it.toSporsmal() }
    )

private fun SoknadsperiodeDTO.toSoknadPeriode(): SoknadPeriode =
    SoknadPeriode(
        fom = fom,
        tom = tom,
        grad = sykmeldingsgrad,
        faktiskGrad = faktiskGrad
    )

fun DeprecatedSykepengesoknadDTO.toSykepengesoknad(): Sykepengesoknad {
    return Sykepengesoknad(
        id = id!!,
        sykmeldingId = sykmeldingId,
        aktorId = aktorId!!,
        soknadstype = type.enumValueOrNull()!!,
        status = status!!.name,
        fom = fom,
        tom = tom,
        opprettet = opprettet!!,
        sendtNav = sendtNav,
        sendtArbeidsgiver = sendtArbeidsgiver,
        arbeidsgiver = arbeidsgiver?.navn,
        arbeidssituasjon = arbeidssituasjon.enumValueOrNull(),
        startSykeforlop = startSyketilfelle,
        sykmeldingSkrevet = sykmeldingSkrevet,
        korrigertAv = korrigertAv,
        korrigerer = korrigerer,
        soknadPerioder = soknadsperioder?.map { it.toSoknadPeriode() },
        sporsmal = sporsmal!!.map { it.toSporsmal() },
        avsendertype = avsendertype.enumValueOrNull(),
        ettersending = ettersending,
        egenmeldtSykmelding = egenmeldtSykmelding,
        orgNummer = arbeidsgiver?.orgnummer,
        harRedusertVenteperiode = harRedusertVenteperiode ?: false,
        merknaderFraSykmelding = merknaderFraSykmelding?.map { Merknad(type = it.type, beskrivelse = it.beskrivelse) }

    )
}

fun SykepengesoknadDTO.toSykepengesoknad(
    aktorId: String
): Sykepengesoknad {
    return Sykepengesoknad(
        id = id,
        sykmeldingId = sykmeldingId,
        aktorId = aktorId,
        soknadstype = type.enumValueOrNull()!!,
        status = status.name,
        fom = fom,
        tom = tom,
        opprettet = opprettet!!,
        sendtNav = sendtNav,
        sendtArbeidsgiver = sendtArbeidsgiver,
        arbeidsgiver = arbeidsgiver?.navn,
        arbeidssituasjon = arbeidssituasjon.enumValueOrNull(),
        startSykeforlop = startSyketilfelle,
        sykmeldingSkrevet = sykmeldingSkrevet,
        korrigertAv = korrigertAv,
        korrigerer = korrigerer,
        soknadPerioder = soknadsperioder?.map { it.toSoknadPeriode() },
        sporsmal = sporsmal!!.map { it.toSporsmal() },
        avsendertype = avsendertype.enumValueOrNull(),
        ettersending = ettersending,
        egenmeldtSykmelding = egenmeldtSykmelding,
        orgNummer = arbeidsgiver?.orgnummer,
        harRedusertVenteperiode = harRedusertVenteperiode ?: false,
        merknaderFraSykmelding = merknaderFraSykmelding?.map { Merknad(type = it.type, beskrivelse = it.beskrivelse) }

    )
}
